src/wcl/doc.h

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cassert>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+namespace wcl {
+
+// Internal state for a doc object. It is intentionally not exposed in the
+// public API and should not be used directly.
+struct doc_state {
+  size_t newline_count = 0;
+  size_t first_width = 0;
+  size_t last_width = 0;
+
+  // Returns the merged state of this and other.
+  doc_state merged(doc_state other) {
+    doc_state merged;
+
+    merged.newline_count = newline_count + other.newline_count;
+
+    merged.last_width = other.last_width;
+    // If other doesn't have any newlines, then the left side needs to be accounted for
+    if (other.newline_count == 0) {
+      merged.last_width += last_width;
+    }
+
+    merged.first_width = first_width;
+    // If the left side doesn't have any newlines, then other side needs to be accounted for
+    if (newline_count == 0) {
+      merged.first_width += other.first_width;
+    }
+
+    return merged;
+  }
+
+  static doc_state from_string(const std::string& str) {
+    doc_state state;
+    for (auto& c : str) {
+      if (c == '\n') {
+        state.last_width = 0;
+        state.newline_count++;
+        continue;
+      }
+
+      if (state.newline_count == 0) {
+        state.first_width++;
+      }
+
+      state.last_width++;
+    }
+    return state;
+  }
+};
+
+class doc_impl_base {
+ protected:
+  // On construction this has to be set
+  size_t size_;
```

Comment:
size can be added to doc_state...if you wanted to be cheecky you could even call it "area" instead of size ;) But probably don't call it area lol.

Reply:
going with `character_count` matches `newline_count` and is more explicit than `size`